rust/src/pop3/pop3.rs

```diff
@@ -0,0 +1,498 @@
+                }
+                Err(SawpError {
+                    kind: SawpErrorKind::Incomplete(sawp::error::Needed::Unknown),
+                }) => {
+                    let consumed = input.len() - start.len();
+                    let needed = start.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(_) => return AppLayerResult::err(),
+            }
+        }
+
+        // Input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn parse_response(&mut self, input: &[u8], flow: *const Flow) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            unsafe {
+                if probe(input) != ALPROTO_POP3 {
+                    // The parser now needs to decide what to do as we are not in sync.
+                    // For this pop3, we'll just try again next time.
+                    return AppLayerResult::ok();
+                }
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.response_gap = false;
+        }
+        let mut start = input;
+        while !start.is_empty() {
+            match POP3_PARSER.parse(start, Direction::ToClient) {
+                Ok((rem, Some(msg))) => {
+                    if let InnerMessage::Response(mut response) = msg.inner {
+                        let tx = if let Some(tx) = self.find_request() {
+                            tx
+                        } else {
+                            // Server sends banner before any requests
+                            let tx = match self.new_tx() {
+                                Some(tx) => tx,
+                                None => return AppLayerResult::err(),
+                            };
+
+                            let tx_id = tx.id();
+                            self.transactions.push_back(tx);
+                            self.get_tx_mut(tx_id - 1).unwrap()
+                        };
+
+                        tx.error_flags |= msg.error_flags;
+                        tx.complete = true;
+
+                        if response.status == sawp_pop3::Status::OK && tx.command.is_some() {
+                            let command = tx.command.as_ref().unwrap();
+                            match &command.keyword {
+                                sawp_pop3::Keyword::STLS => {
+                                    unsafe {
+                                        AppLayerRequestProtocolTLSUpgrade(flow);
+                                    };
+                                }
+                                sawp_pop3::Keyword::RETR => {
+                                    // Don't hold onto the whole email body
+
+                                    // TODO: pass off to mime parser
```

Comment:
Is the idea here to use the mime parsing Suricata already has? Or are you thinking about some other parser?

Reusing the suri code would mean it's also trivial to support the `email.*` keywords we're adding.

Reply:
That plan was to use the Suricata mime parser. Didn't want to figure it out for this PR, but could make a ticket to do it after.